kube/src/service/mod.rs

```diff
@@ -104,22 +103,47 @@ impl TryFrom<Config> for Service {
             .map_response(maybe_decompress)
             .into_inner();
 
-        let https: HttpsConnector<_> = config.try_into()?;
-        let mut connector = TimeoutConnector::new(https);
-        if let Some(timeout) = timeout {
-            // reqwest's timeout is applied from when the request stars connecting until
-            // the response body has finished.
-            // Setting both connect and read timeout should be close enough.
-            connector.set_connect_timeout(Some(timeout));
-            // Timeout for reading the response.
-            connector.set_read_timeout(Some(timeout));
-        }
-        let client: HyperClient<_, Body> = HyperClient::builder().build(connector);
+        let with_timeout = layer::layer_fn(|c| {
+            let mut connector = TimeoutConnector::new(c);
+            if let Some(timeout) = timeout {
+                // reqwest's timeout is applied from when the request stars connecting until
+                // the response body has finished.
+                // Setting both connect and read timeout should be close enough.
+                connector.set_connect_timeout(Some(timeout));
+                // Timeout for reading the response.
+                connector.set_read_timeout(Some(timeout));
+            }
+            connector
+        });
+
+        #[cfg(feature = "native-tls")]
+        let tls = connector::tls_connector(config.identity, config.root_cert, config.accept_invalid_certs)?;
+        #[cfg(feature = "rustls-tls")]
+        let tls = std::sync::Arc::new(connector::tls_config(
+            config.identity,
+            config.root_cert,
+            config.accept_invalid_certs,
+        )?);
+
+        #[cfg(not(feature = "proxy"))]
+        let conn = connector::https_connector(tls);
+        #[cfg(feature = "proxy")]
+        let conn = connector::proxy_connector(
+            tls,
+            if let Some(url) = &config.proxy_url {
+                Some(url.parse::<http::uri::Uri>()?)
+            } else {
+                None
+            },
```

Comment:
Thought this could be nicer with `Option::map`, but think error handling screws it up into

```rust
config.proxy_url.map(|url| url.parse::<http::uri::Uri>()).transpose()?
```
which is probably a bit convoluted.

Reply:
Thanks! I did not know [`transpose`](https://doc.rust-lang.org/std/option/enum.Option.html#method.transpose) existed! :exploding_head:

The type can be inferred, so it can be:

```rust
let conn = connector::proxy_connector(
    tls,
    config.proxy_url.map(|s| s.parse()).transpose()?
);
```
`config.proxy_url.map(str::parse).transpose()?` doesn't work.